### ISSUE TITLE
tokio::spawn fails to work inside compat'd code

### ISSUE THREAD

**Issue by djmarcin** [LucioFranco/tokio-compat-02#10]
I'm using this crate to interop with tonic, which is still on 0.2. When tonic calls into my code to handle an RPC, it appears to do so within a 0.2 context, which makes sense. However, if I try to use `tokio::spawn` within that code, it uses my 0.3 version of spawn, which panics because it can't find the runtime:
```
Thread tokio-runtime-worker panicked at 'must be called from the context of Tokio runtime configured with either `basic_scheduler` or `threaded_scheduler`'
```

**Comment by Darksonn** [comment]
Check out the hyper server example in the 02 compat crate.

**Comment by NicholasGorski** [comment]
The server example doesn't cover this. I ran into this as well.

This crate fundamentally cannot fix this. When you reference `tokio::spawn`, you're referencing a function scoped to a version of the code. One can either: reference the matching 0.2 version of the code so it finds the 0.2 runtime, or explicitly keep a handle to tokio 0.3/1.0 and spawn through the handle.

The former would require adding tokio_02 matching this crate to your own Cargo, e.g.:

```
tokio_02 = {version = "0.2.22", package = "tokio", features = ["rt-threaded"]}
```

Now you can use the spawn function from that version, e.g. `tokio_02::task::spawn(...)`. This crate creates a Tokio 0.2 runtime that's single-threaded, though, so adding another task has limited usefulness in multithreading. (You can of course fork the repo like I did and tweak it to your liking.)

To spawn in tokio 0.3/1.0 from 0.2, you need to get a handle to the runtime beforehand. For example, while in a tokio 0.3/1.0 context:

```
let handle = tokio::runtime::Handle::current();
```

This value is cloneable. Later you can invoke `handle.spawn()` and get the same behavior as if you called `tokio::spawn` from your desired runtime. (`tokio::spawn` is more or less just this.) For example:

```
struct MyService {
    handle: tokio::runtime::Handle
}

impl MyService {
    pub fn new() -> Self {
        Self { handle: tokio::runtime::Handle::current() }
    }
    
    async fn foo_compat_02(&self, ..) {
        self.handle.spawn(self.foo(..)).await.unwrap()
    }
    
    async fn foo(&self, ..) {
        // run in the "natural" tokio runtime, no longer in 0.2.
    }
}
```

I have found this pattern more useful, as it minimizes the time spent in the special compat runtime.

Hope that helps!

**Comment by Darksonn** [comment]
@NicholasGorski Can you elaborate? There appears to be some confusion here because the server example _does_ cover this.

> This crate fundamentally cannot fix this. When you reference `tokio::spawn`, you're referencing a function scoped to a version of the code. One can either: reference the matching 0.2 version of the code so it finds the 0.2 runtime, or explicitly keep a handle to tokio 0.3/1.0 and spawn through the handle.

As long as you are still inside the 0.3/1 runtime context, you can use the 0.3/1 `tokio::spawn` method without issue. Then, when you spawn, you can mark the newly spawned task with `.compat()` to make it enter the 0.2 context as well. The hyper example does this in the following manner:

https://github.com/LucioFranco/tokio-compat-02/blob/427c72cf9290f852feae84846b440b5ca71e57ec/examples/hyper_server.rs#L53-L62

Generally the place where you run into trouble is if you call the 0.2 `tokio::spawn` method, because then you are _outside_ the 0.3/1 runtime context. So don't do that. Always run your tasks on the 0.3/1 runtime, using `.compat()` on the newly spawned task.

**Comment by NicholasGorski** [comment]
Right, everything you said is correct. What @djmarcin is asking is when "as long as you are still inside the 0.3/1 runtime context" does not hold true.

For example, warp and tonic will (abstractly) accept handler functions, which when invoked return a future to run. These libraries internally call 0.2 `tokio::spawn`, so people are running `serve(my_api_handlers).compat().await` to get things working. This way warp and tonic can be started from an 0.3/1 context and be executed inside the 0.2 runtime.

What happens is warp/tonic invokes the handler function, gets a future, then calls `<0.2>::tokio::task::spawn(the_handler)`. Now the handler is running in the 0.2 context as well. If this handler wants to do anything tokio, it must do it using the 0.2 functions. This is what OP is referring to with: "if I try to use tokio::spawn within that code, it uses my 0.3 version of spawn, which panics because it can't find the runtime".

To "get back" to the 0.3 runtime, one has to effectively re-run the task back on the 0.3/1 runtime using a `Handle`. This crate makes it very easy to get a future to run in an 0.2. runtime, but understandably cannot magically know when the user would like to (safely) get back to the 0.3/1 runtime.

**Comment by Darksonn** [comment]
Ideally you would configure the server to use a different spawn function. If the server does not provide any way to do that, I suppose that you are right that you do need a `Handle`.
